### PR TITLE
ref(tx-processor): Apply transaction rules after UUID scrubbing

### PR DESCRIPTION
**Pull request by iker-barriocanal**
Applying transaction rules before UUID scrubbing makes it impossible to track what rules were applied. It's interesting to know what rules are useful to reduce the high cardinality, so that we can use them again over other rules that are less interesting or provide no value. Applying these rules after UUID scrubbing leaves no room for further modification of the transaction name, so the resulting transaction name is guaranteed to be the direct output of a rule.

This is the first step in the process of discarding the rules that provide no value and prioritizing those that do.

Ref: https://github.com/getsentry/team-ingest/issues/14.

### CODE REVIEW

File: relay-general/src/store/transactions/processor.rs [hunk 1]
```diff
@@ -341,28 +341,32 @@ impl Processor for TransactionsProcessor<'_> {
...
-        ) && self.name_config.scrub_identifiers
-        {
-            let changed = scrub_identifiers(&mut event.transaction)?;
-            if changed && self.name_config.mark_scrubbed_as_sanitized {
-                let source = &mut event
-                    .transaction_info
-                    .get_or_insert_with(Default::default)
-                    .source;
-                source.set_value(Some(TransactionSource::Sanitized));
+        if self.name_config.scrub_identifiers {
```
Comment (jjbayer) [inline relay-general/src/store/transactions/processor.rs:350]: nit: Should we rename the `scrub_identifiers` flag in config to `normalize`?

I've been thinking about naming for a while, I think we need unambiguous names to refer to the different parts of transaction name handling but haven't been able to come up with good ones.

File: relay-general/src/store/transactions/processor.rs [hunk 2]
```diff
@@ -341,28 +341,32 @@ impl Processor for TransactionsProcessor<'_> {
...
+                    // TODO(iker): we also mark transaction as sanitized after
+                    // applying renaming rules. Once we remove this feature
+                    // flag, relay should only mark transactions as sanitized
+                    // once.
+                    event
+                        .transaction_info
+                        .get_or_insert_with(Default::default)
+                        .source
+                        .set_value(Some(TransactionSource::Sanitized));
+                }
```
Comment (jjbayer) [inline relay-general/src/store/transactions/processor.rs]: It might not matter in practice, but I would put this block _after_ `apply_transaction_rename_rule`. Basically

1. Apply sanitation step 1
2. Apply sanitation step 2
3. Mark as sanitized

Right now the order is

1. Apply sanitation step 1
1. Mark as sanitized
1. Apply sanitation step 2

which feels unnatural to me.

File: relay-general/src/store/transactions/processor.rs (hunk 1 above)
Comment (iker-barriocanal) [reply relay-general/src/store/transactions/processor.rs:350]: I agree the names could be better, but I don't have any suggestions (besides being very explicit and setting extremely long names, which doesn't really improve things I guess). Since I expect to remove these feature flags soonish, I didn't rename them. That said, happy to refactor the names if you come up with a good suggestion.

File: relay-general/src/store/transactions/processor.rs (hunk 2 above)
Comment (iker-barriocanal) [reply relay-general/src/store/transactions/processor.rs]: Docs of `mark_scrubbed_as_sanitized` are as follows:

> True if transaction names scrubbed by regex patterns should be marked as `TransactionSource::Sanitized`.
>
> Transaction names modified by clusterer rules are always marked as such.

So the code follows these steps:

1. Apply sanitation step 1.
2. Mark as sanitized, resulting from step 1.
3. Apply sanitation step 2 (which also marks as sanitized, so no `4` needed).

I think 1 and 2 belong together. I agree that this could be improved, and that's why I added the `TODO`, but I wanted to minimize refactoring as much as possible until we have more information on how this looks and what we do with the feature flags..

Comment (jjbayer) [reply relay-general/src/store/transactions/processor.rs]: > True if transaction names scrubbed by regex patterns should be marked as TransactionSource::Sanitized.

That doc was written before you removed the `if changed` condition though (by which it actually belonged to pattern-based matching). By removing that condition, the real consequence of `mark_scrubbed_as_sanitized` is that _every_ `url` transaction will be marked as sanitized as long as the `normalize-transaction-names` feature flag is on.

File: relay-general/src/store/transactions/processor.rs [hunk 3]
```diff
@@ -57,7 +57,13 @@ impl<'r> TransactionsProcessor<'r> {
...
                 if let Some((rule, result)) = result {
                     if *transaction != result {
-                        meta.set_original_value(Some(transaction.clone()));
+                        // If another rule was applied before, we don't want to
+                        // rename the transaction name to keep the original one.
+                        // We do want to continue adding remarks though, in
+                        // order to keep track of all rules applied.
+                        if meta.original_value().is_none() {
+                            meta.set_original_value(Some(transaction.clone()));
+                        }
```
Comment (iker-barriocanal) [inline relay-general/src/store/transactions/processor.rs:66]: I tried to do this with `get_or_insert_with`, but I couldn't figure out how. Happy to change it.

File: relay-general/src/store/transactions/processor.rs [hunk 4]
```diff
@@ -1675,7 +1684,7 @@ mod tests {
         let json = r#"
         {
             "type": "transaction",
-            "transaction": "/foo/2fd4e1c67a2d28fced849ee1bb76e7391b93eb12/user/123/0/",
+            "transaction": "/foo/rule-target/user/123/0/",
```
Comment (iker-barriocanal) [inline relay-general/src/store/transactions/processor.rs:1686]: The reason for these changes is to avoid UUID scrubbing rules running first in all fields. They are already running for `123` so that is good enough.

File: relay-general/src/store/transactions/processor.rs [hunk 5]
```diff
@@ -356,13 +356,22 @@ impl Processor for TransactionsProcessor<'_> {
...
+                    // once.
                     event
                         .transaction_info
                         .get_or_insert_with(Default::default)
                         .source
                         .set_value(Some(TransactionSource::Sanitized));
                 }
             }
+
+            self.apply_transaction_rename_rule(
```
Comment (jjbayer) [inline relay-general/src/store/transactions/processor.rs:360]: When `mark_scrubbed_as_sanitized` is enabled (which it will be soon), the rules matching on `source: url` will fail to match, so `apply_transaction_rename_rule` will essentially become dead code. Maybe the previously failing tests already reflected this, I did not check in detail.

This brings me back to the point that we should "mark as sanitized" after applying the rules.

**Review by jjbayer — CHANGES_REQUESTED** [review]
See https://github.com/getsentry/relay/pull/1964#discussion_r1148924220.

File: relay-general/src/store/transactions/processor.rs (hunk 5 above)
Comment (iker-barriocanal) [reply relay-general/src/store/transactions/processor.rs:360]: This is a good catch! It should now be addressed, including a test to avoid this from happening in the future.

File: relay-general/src/store/transactions/processor.rs (hunk 3 above)
Comment (jjbayer) [reply relay-general/src/store/transactions/processor.rs:66]: I think this is nice and explicit.

File: relay-general/src/store/transactions/processor.rs [hunk 6]
```diff
@@ -342,26 +348,30 @@ impl Processor for TransactionsProcessor<'_> {
...
+            }
+
+            self.apply_transaction_rename_rule(
+                &mut event.transaction,
+                event.transaction_info.value_mut(),
+            )?;
+
+            if !matches!(
+                event.get_transaction_source(),
+                &TransactionSource::Sanitized
```
Comment (jjbayer) [inline relay-general/src/store/transactions/processor.rs]: Wouldn't this set `source=sanitized` for every source (`route`, `custom`, etc.). I think we need to match on `TransactionSource::Url` here explicitly.

**Review by jjbayer — CHANGES_REQUESTED** [review]

File: relay-general/src/store/transactions/processor.rs (hunk 6 above)
Comment (iker-barriocanal) [reply relay-general/src/store/transactions/processor.rs]: Heh, I think I got confused _again_ from the code structure. Let's see when we can remove the feature flags here and make the code a bit easier to follow.

Thanks for putting in the effort!

File: relay-general/src/store/transactions/processor.rs [hunk 7]
```diff
@@ -362,10 +362,8 @@ impl Processor for TransactionsProcessor<'_> {
                 event.transaction_info.value_mut(),
             )?;
 
-            if !matches!(
-                event.get_transaction_source(),
-                &TransactionSource::Sanitized
-            ) && self.name_config.mark_scrubbed_as_sanitized
+            if matches!(event.get_transaction_source(), &TransactionSource::Url)
+                && self.name_config.mark_scrubbed_as_sanitized
```
Comment (jjbayer) [inline relay-general/src/store/transactions/processor.rs:366]: As discussed offline, we might want to add an an additional condition which is `scrub_identifiers_changed_something OR rename_rules.len() > 0`, to prevent projects that do not have any rules yet from marking everything as sanitized. But that is actually out of scope for this PR.

**Review by jjbayer — APPROVED** [review]
